src/main/java/com/aws/greengrass/disk/spool/DiskSpoolDAO.java

```diff
@@ -71,22 +82,22 @@ public DiskSpoolDAO(NucleusPaths paths) throws IOException {
      */
     @SuppressWarnings("PMD.AvoidCatchingGenericException")
     public Iterable<Long> getAllSpoolMessageIds() throws SQLException {
-        List<Long> currentIds;
-        String query = "SELECT message_id FROM spooler;";
-        try (Connection conn = getDbInstance();
-             PreparedStatement pstmt = conn.prepareStatement(query);
-             ResultSet rs = RetryUtils.runWithRetry(sqlStatementRetryConfig, pstmt::executeQuery,
-                     "get-all-spool-message-ids", logger)) {
-            currentIds = getIdsFromRs(rs);
-        } catch (SQLException e) {
-            checkAndHandleCorruption(e);
-            throw e;
-        } catch (Exception e) {
-            throw new SQLException("Failed to get Spool Message IDs", e);
+        synchronized (dbConnection) {
```

Comment:
instead of synchronizing on the connection (which is an object that we know will be created and destroyed over time), use a [ReentrantReadWriteLock](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html). Use the read lock when using the connection. Use the write lock when creating or closing the connection.

Reply:
And use [LockScope](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/a368ad647bccd44f9689226750afd87f60a84e86/src/main/java/com/aws/greengrass/util/LockScope.java#L10) to ensure the lock is acquired and released properly.